include/wil/cppwinrt_notifiable_module_lock.h

```diff
@@ -0,0 +1,103 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+//! @file
+//! Utilities for implementing OOP COM server with cppwinrt only
+
+#ifndef __WIL_CPPWINRT_NOTIFIABLE_MODULE_LOCK_DEFINED
+#define __WIL_CPPWINRT_NOTIFIABLE_MODULE_LOCK_DEFINED
+
+#if _MSVC_LANG >= 201703L
+
+#include <atomic>
+#include <cstdint>
+
+// Ligh up cppwinrt custom module lock
+#define WINRT_CUSTOM_MODULE_LOCK

```

Comment:
What is the effect if this header is included after `<cppwinrt/base.h>`? What is the effect if some translation units include this header, but others don't?

For the first question, an `#ifdef` w/ `#error` would likely add value if that silently fails. I'm not sure we can do anything about the second; C++/WinRT would likely need to be the one with an ODR check.

Reply:
Maybe I'm answering my own question, but hopefully the linker would complain about the two different versions of `winrt::get_module_lock`. Hopefully...

Reply:
Suggest adding `#ifdef WINRT_BASE_H` as a more explicit guard/error against this case. Unfortunately I don't think it can catch everything (for example TUs which include `winrt/base.h` but never `wil/cppwinrt_notifiable_module_lock.h`), but hopefully people making use of this functionality define `WINRT_CUSTOM_MODULE_LOCK` at the project level.

What if we simply ommitted defining `WINRT_CUSTOM_MODULE_LOCK` to encourage users to define it at the project level?

Reply:
> If you define another module lock after including `winrt/base.h`, the macro itself has no effect as the whole `winrt/base.h` is guarded by `#pragma once`.

Sorry, maybe I wasn't clear. From what I can gather, this header must be included _before_ any C++/WinRT headers and must be included in any translation unit that also includes any C++/WinRT headers. I wasn't trying to imply that including `<cppwinrt/base.h>` multiple times was somehow problematic or desired.

> Subsequent redefinition of `winrt::get_module_lock` get various kind of compile/linker error.

I would definitely expect that including `<cppwinrt/base.h>` before this header would result in compilation errors due to the conflicting definitions of `cppwinrt::get_module_lock` in the same TU. It's probably a nasty non-obvious compilation error that's likely not simple to diagnose, but at least that should fail consistently. Do you also get a failure from the linker error if another TU includes C++/WinRT, but not this header? You'd still have two conflicting definitions, however it's my understanding that the linker is not required to emit a diagnostic here (though if `link.exe` currently does that makes me feel a little better).

> What if we simply ommitted defining `WINRT_CUSTOM_MODULE_LOCK` to encourage users to define it at the project level?

Pending whether or not the linker fails for this scenario, that might be an option. We'd want to add something like:
```c++
#ifndef WINRT_CUSTOM_MODULE_LOCK
#error You must define 'WINRT_CUSTOM_MODULE_LOCK' at the project level
#endif
```
Though it would still be great if C++/WinRT had its own ODR check here

Reply:
> What if we simply ommitted defining WINRT_CUSTOM_MODULE_LOCK to encourage users to define it at the project level?

That works. Current implementation is merely a "least resistance path" that does everything for you. I did think of making it opt-in. That is to say, define another macro, say, `WIL_USE_DEFAULT_NOTIFIABLE_MODULE_LOCK` which wire everything up.

I'm fine with both opt-out or opt-in. But perhaps reviewers/MS folks would prefer opt-in?

Reply:
Although in practice I'd expect the `notifiable_module_lock` header to be included in `pch.h` and thus it's very rare that there'd be ODR issues, I added the ODR guard just to be safe. Note that API wise, this means the least resistance path needs to be enabled with defining `WINRT_CUSTOM_MODULE_LOCK` + including `notifiable_module_lock` header. For the advanced scenario where user want to use notifiable module lock and defining his own lock, he'd need to also define `WIL_CPPWINRT_COM_SERVER_CUSTOM_MODULE_LOCK`, as shown in the test.

Reply:
I have also added the following guard
```
#ifndef WINRT_BASE_H
#error You must include this header before including any winrt header
#endif
```